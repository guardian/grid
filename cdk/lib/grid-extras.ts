import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import {
	GuStack,
} from '@guardian/cdk/lib/constructs/core';
import type { App } from 'aws-cdk-lib';
import {aws_events, aws_lambda, aws_ssm, Duration, SecretValue} from 'aws-cdk-lib';
import {
	ENV_VAR_BATCH_SIZE,
	ENV_VAR_MEDIA_API_HOSTNAME,
	ENV_VAR_MEDIA_API_KEY,
} from '../../reaper/src/envVariables';

const ALARM_SNS_TOPIC_NAME = 'Cloudwatch-Alerts';

interface GridExtrasProps extends GuStackProps {
	batchSize: number;
}
export class GridExtras extends GuStack {
	constructor(scope: App, id: string, props: GridExtrasProps) {
		super(scope, id, props);

		const reaperApp = 'grid-reaper';
		new GuScheduledLambda(this, 'ReaperLambda', {
			app: reaperApp,
			functionName: `grid-reaper-lambda-${this.stage}`,
			runtime: aws_lambda.Runtime.NODEJS_16_X, // upgrade when .nvmrc is updated
			handler: 'index.handler',
			environment: {
				[ENV_VAR_BATCH_SIZE]: props.batchSize.toString(),
				[ENV_VAR_MEDIA_API_HOSTNAME]: aws_ssm.StringParameter.valueForStringParameter(
          this,
          `/${this.stage}/${props.stack}/${reaperApp}/mediaApiHostname`
        ),
				[ENV_VAR_MEDIA_API_KEY]: SecretValue.ssmSecure(
          `/${this.stage}/${props.stack}/${reaperApp}/mediaApiKey`,

        ).toString()
			},
			monitoringConfiguration: {
				toleratedErrorPercentage: 0,
				snsTopicName: ALARM_SNS_TOPIC_NAME,
				okAction: true,
			},
			fileName: 'grid-reaper.zip',
			rules: [
				{
					// every 15mins at 1000 each time = 96,000 per day
					schedule: aws_events.Schedule.rate(Duration.minutes(15)),
					description:
						'Run every 15 mins to ensure we keep on top of the backlog of images to be reaped',
				},
			],
		});
	}
}
