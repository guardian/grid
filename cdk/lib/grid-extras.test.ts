import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GridExtras } from './grid-extras';

describe('The GridExtras stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new GridExtras(app, 'GridExtras', {
			stack: 'media-service',
			stage: 'TEST',
			batchSize: 99, //arbitrary number under 1000
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
