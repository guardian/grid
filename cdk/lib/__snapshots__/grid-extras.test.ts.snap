// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GridExtras stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueTESTmediaservicegridreapermediaApiHostnameC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/TEST/media-service/grid-reaper/mediaApiHostname",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ReaperLambda78A3B842": {
      "DependsOn": [
        "ReaperLambdaServiceRoleDefaultPolicyE0B067D6",
        "ReaperLambdaServiceRole40E0308B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "media-service/TEST/grid-reaper/grid-reaper.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "grid-reaper",
            "BATCH_SIZE": "99",
            "MEDIA_API_HOSTNAME": {
              "Ref": "SsmParameterValueTESTmediaservicegridreapermediaApiHostnameC96584B6F00A464EAD1953AFF4B05118Parameter",
            },
            "MEDIA_API_KEY": "{{resolve:ssm-secure:/TEST/media-service/grid-reaper/mediaApiKey}}",
            "STACK": "media-service",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "grid-reaper-lambda-TEST",
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ReaperLambdaServiceRole40E0308B",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "grid-reaper",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/grid",
          },
          {
            "Key": "Stack",
            "Value": "media-service",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ReaperLambdaErrorPercentageAlarmForLambda53A6409C": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":Cloudwatch-Alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "ReaperLambda78A3B842",
              },
              " exceeded 0% error rate",
            ],
          ],
        },
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "High error % from ",
              {
                "Ref": "ReaperLambda78A3B842",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "ReaperLambda78A3B842",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "ReaperLambda78A3B842",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "ReaperLambda78A3B842",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "OKActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":Cloudwatch-Alerts",
              ],
            ],
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ReaperLambdaReaperLambdarate15minutes039B86344": {
      "Properties": {
        "Description": "Run every 15 mins to ensure we keep on top of the backlog of images to be reaped",
        "ScheduleExpression": "rate(15 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ReaperLambda78A3B842",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ReaperLambdaReaperLambdarate15minutes0AllowEventRuleGridExtrasReaperLambda1164EFA938AF1E54": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ReaperLambda78A3B842",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ReaperLambdaReaperLambdarate15minutes039B86344",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ReaperLambdaServiceRole40E0308B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "grid-reaper",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/grid",
          },
          {
            "Key": "Stack",
            "Value": "media-service",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ReaperLambdaServiceRoleDefaultPolicyE0B067D6": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/media-service/TEST/grid-reaper/grid-reaper.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/media-service/grid-reaper",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/media-service/grid-reaper/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ReaperLambdaServiceRoleDefaultPolicyE0B067D6",
        "Roles": [
          {
            "Ref": "ReaperLambdaServiceRole40E0308B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
